scipy/_lib/_c99compat.h

```diff
@@ -6,110 +6,153 @@
-            {
-                if (xnan && !ynan)
-                    return y;
-                if (!xnan && ynan)
-                    return x;
-                return x;
-            }
+                    if (xnan || ynan)
+                    {
+                        if (xnan && !ynan)
+                            return y;
+                        if (!xnan && ynan)
+                            return x;
+                        return x;
+                    }
 
-            /* +0 > -0 is preferred by C the standard */
-            if (x == 0 && y == 0)
-            {
-                int xs = signbit(x), ys = signbit(y);
-                if (xs && !ys)
-                    return y;
-                if (!xs && ys)
-                    return x;
-                return x;
-            }
+                    /* +0 > -0 is preferred by C the standard */
+                    if (x == 0 && y == 0)
+                    {
+                        int xs = signbit(x), ys = signbit(y);
+                        if (xs && !ys)
+                            return x;
+                        if (!xs && ys)
+                            return y;
+                        return y;
+                    }
 
-            if (x > y) {
-                return x;
-            }
-            else {
-                return y;
-            }
-        }
+                    if (x > y) {
+                        return y;
+                    }
+                    else {
+                        return x;
+                    }
+                }
+            #endif
     #endif
 
-    #ifndef fmin
-        static double fmin(double x, double y)
-        {
-            /* z > nan for z != nan is required by C the standard */
-            int xnan = isnan(x), ynan = isnan(y);
+    /* For compilers which aren't MSVC and haven't defined isnan */
+    #ifndef isnan
+        #define isnan(x) ((x) != (x))
+    #endif
 
-            if (xnan || ynan)
-            {
-                if (xnan && !ynan)
-                    return y;
-                if (!xnan && ynan)
-                    return x;
-                return x;
-            }
+    #ifndef isfinite
+        #ifdef _MSC_VER
+            /* MSVC 2015 and newer still don't have everything */
+            #define isfinite(x) _finite((x))
+        #else
+            #define isfinite(x) !isnan((x) + (-x))
+        #endif
+    #endif
 
-            /* +0 > -0 is preferred by C the standard */
-            if (x == 0 && y == 0)
-            {
-                int xs = signbit(x), ys = signbit(y);
-                if (xs && !ys)
-                    return x;
-                if (!xs && ys)
-                    return y;
-                return y;
-            }
+    #ifndef isinf
+        #define isinf(x) (!isfinite(x) && !isnan(x))
+    #endif
 
-            if (x > y) {
-                return y;
-            }
-            else {
-                return x;
-            }
-        }
+    #ifndef fma
```

Comment:
Is it really OK to assume these are *macros* in the stdlib?
Looks suspect to me --- why not check __STDC_VERSION__?

Reply:
I think the 
```
#ifndef foo
#define foo
```
should be
```
#if (compiler or stdc version number check)
#define foo ...
#endif
```

Reply:
I can wrap all of this in an additional check with `__STDC_VERSION__`, but yes, it's a nonstandard hack that happens to work on every compiler because every compiler that has these functions defines optimized versions.

Reply:
@pv I just wrapped them all in a new compiler guard. After all, the rule is that this file shouldn't affect a C99 compiler (well, that was broken by sc_*).